compiler/transforms/reuse_memref_space.py

```diff
@@ -0,0 +1,254 @@
+from collections.abc import Sequence
+
+from xdsl.context import MLContext
+from xdsl.dialects import affine, arith, builtin, memref, scf
+from xdsl.dialects.builtin import IndexType
+from xdsl.ir import Block, Operation, OpResult, SSAValue
+from xdsl.ir.affine import AffineConstantExpr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.rewriter import InsertPoint
+
+
+class MoveMemrefAllocations(RewritePattern):
+    """
+    This class represents a rewrite pattern for moving memref allocations outside
+    a double for-loop. This is possible when each loop a new memref-space is allocated with
+    identical sizes and the space is not used outside the loop.
```

Comment:
A few things come to mind reading this:
* Why does this mention a double-for loop? Does it not work for double loops? Maybe we should state that it works for `nested loops`?
* When talking about memref-space, do you mean the space associated with the storage of a memref? If that is the case I would just refer to it as `the memref`. Now it is a bit misleading, since memrefs also have a memory space associated with them, which does not mean the space, but rather a level in memory, like "L3" or "LLC".